File: killrvideo-service-users/src/test/java/com/killrvideo/service/user/grpc/UserManagementServiceGrpcMapperTest.java
package com.killrvideo.service.user.grpc;

import com.killrvideo.service.user.dto.User;
import killrvideo.user_management.UserManagementServiceOuterClass.*;
import killrvideo.user_management.events.UserManagementEvents;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

import static com.killrvideo.utils.GrpcMappingUtils.uuidToUuid;
import static java.util.Collections.singletonList;
import static org.junit.jupiter.api.Assertions.*;

class UserManagementServiceGrpcMapperTest {
    private UserManagementServiceGrpcMapper mapper = new UserManagementServiceGrpcMapper();

    @Test
    public void testMapUserRequest2User() {
        User u = user();
        CreateUserRequest request = CreateUserRequest.newBuilder()
                .setEmail(u.getEmail())
                .setFirstName(u.getFirstname())
                .setLastName(u.getLastname())
                .setUserId(uuidToUuid(u.getUserid()))
                .build();

        User user = mapper.mapUserRequest2User(request);
        assertEquals(u.getEmail(), user.getEmail());
        assertEquals(u.getFirstname(), user.getFirstname());
        assertEquals(u.getLastname(), user.getLastname());
        assertEquals(u.getUserid(), user.getUserid());
    }

    @Test
    public void testMapResponseVerifyCredentials() {
        UUID userid = UUID.randomUUID();
        VerifyCredentialsResponse response = mapper.mapResponseVerifyCredentials(userid);
        assertEquals(userid.toString(), response.getUserId().getValue());
    }

    @Test
    public void testCreateUserCreatedEvent() {
        User u = user();
        UserManagementEvents.UserCreated event = mapper.createUserCreatedEvent(u);
        assertEquals(u.getEmail(), event.getEmail());
        assertEquals(u.getFirstname(), event.getFirstName());
        assertEquals(u.getLastname(), event.getLastName());
        assertEquals(u.getUserid().toString(), event.getUserId().getValue());
    }

    @Test
    public void testMapToListOfUserid() {
        UUID userid1 = UUID.randomUUID();
        UUID userid2 = UUID.randomUUID();
        GetUserProfileRequest request = GetUserProfileRequest.newBuilder()
                .addUserIds(uuidToUuid(userid1))
                .addUserIds(uuidToUuid(userid2))
                .build();

        List<UUID> userids = mapper.mapToListOfUserid(request);
        assertEquals(2, userids.size());
    }

    @Test
    public void testBuildGetUserProfileResponse() {
        User u = user();
        GetUserProfileResponse response = mapper.buildGetUserProfileResponse(singletonList(u));
        assertEquals(1, response.getProfilesCount());

        UserProfile profile = response.getProfiles(0);

        assertEquals(u.getEmail(), profile.getEmail());
        assertEquals(u.getFirstname(), profile.getFirstName());
        assertEquals(u.getLastname(), profile.getLastName());
        assertEquals(u.getUserid().toString(), profile.getUserId().getValue());
    }

    private User user() {
        UUID userid = UUID.randomUUID();
        String email = "joe@gmail.com";
        String firstName = "first name";
        String lastName = "last name";
        User user = new User();
        user.setUserid(userid);
        user.setEmail(email);
        user.setFirstname(firstName);
        user.setLastname(lastName);
        user.setCreatedDate(Instant.now());
        return user;
    }
}

File: killrvideo-service-users/src/test/kotlin/com/killrvideo/service/user/grpc/UserManagementServiceGrpcMapperTest.kt
package com.killrvideo.service.user.grpc

import com.killrvideo.service.user.dto.User
import com.killrvideo.service.user.grpc.UserManagementServiceGrpcMapper.CreateUserRequestExtensions.parse
import com.killrvideo.utils.GrpcMappingUtils.uuidToUuid
import killrvideo.user_management.createUserRequest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.time.Instant
import java.util.*

internal class UserManagementServiceGrpcMapperTest {
    private val mapper = UserManagementServiceGrpcMapper()
    @Test
    fun testMapUserRequest2User() {
        val u: User = user()
        val request = createUserRequest {
            email = u.email!!
            firstName = u.firstname!!
            lastName = u.lastname!!
            userId = uuidToUuid(u.userid!!)
        }
        val (user, _) = request.parse()
        assertEquals(u.email, user.email)
        assertEquals(u.firstname, user.firstname)
        assertEquals(u.lastname, user.lastname)
        assertEquals(u.userid, user.userid)
    }

    @Test
    fun testMapResponseVerifyCredentials() {
        val userid = UUID.randomUUID()
        val response = mapper.mapResponseVerifyCredentials(userid)
        assertEquals(userid.toString(), response.userId.value)
    }

    @Test
    fun testCreateUserCreatedEvent() {
        val u = user()
        val event = mapper.createUserCreatedEvent(u)
        assertEquals(u.email, event.email)
        assertEquals(u.firstname, event.firstName)
        assertEquals(u.lastname, event.lastName)
        assertEquals(u.userid.toString(), event.userId.value)
    }

    @Test
    fun testBuildGetUserProfileResponse() {
        val u = user()
        val response = mapper.buildGetUserProfileResponse(listOf(u))
        assertEquals(1, response.profilesCount)
        val profile = response.getProfiles(0)
        assertEquals(u.email, profile.email)
        assertEquals(u.firstname, profile.firstName)
        assertEquals(u.lastname, profile.lastName)
        assertEquals(u.userid.toString(), profile.userId.value)
    }

    private fun user(): User {
        val userid = UUID.randomUUID()
        val email = "joe@gmail.com"
        val firstName = "first name"
        val lastName = "last name"
        return User(
            userid = userid,
            email = email,
            firstname = firstName,
            lastname = lastName,
            createdDate = Instant.now()
        )
    }
}
